src/shared/ui/tag-box.tsx

```diff
@@ -0,0 +1,138 @@
+import { Separator } from "./separator";
+
+export type TagBoxHandle = {
+  set: (newTags: string[]) => void;
+};
+
+type ValidationPattern = {
+  message: string;
+  regex: RegExp;
+};
+
+export interface TagBoxProps
+  extends React.InputHTMLAttributes<HTMLInputElement> {
+  onTagsChange?: (tags: string[]) => void;
+  initialTags?: string[];
+  patterns?: ValidationPattern[];
+}
+
+const TagBox = forwardRef<TagBoxHandle, TagBoxProps>(
+  (
+    {
+      className,
+      initialTags = [],
+      onTagsChange,
+      patterns = [],
+      placeholder = "Type and press Enter",
+    },
+    ref
+  ) => {
+    const [tags, setTags] = useState<string[]>(initialTags);
+    const [inputValue, setInputValue] = useState("");
+    const [validationError, setValidationError] = useState<string | null>(null);
+
+    useImperativeHandle(ref, () => ({
+      set: (newTags: string[]) => setTags(newTags),
+    }));
+
+    function validateTag(tag: string): string | null {
+      for (const { message, regex } of patterns) {
+        if (!regex.test(tag)) return message;
+      }
+      return null;
+    }
+
+    function addInputAsTag() {
+      const tag = inputValue.trim();
+      if (isEmptyString(tag.trim())) return;
+
+      const error = validateTag(tag);
+      if (isPopulatedString(error)) {
+        setValidationError(error);
+        return;
+      }
+
+      const updatedTags = mergeUniqueStrings(tags, [tag]);
+      setTags(updatedTags);
+      setInputValue("");
+      onTagsChange?.(updatedTags);
+    }
+
+    const removeTag = (tagToRemove: string) => {
+      const updatedTags = tags.filter((tag) => tag !== tagToRemove);
+      setTags(updatedTags);
+      onTagsChange?.(updatedTags);
+    };
+
+    const handleKeyDown = (e: React.KeyboardEvent<HTMLInputElement>) => {
+      if (e.key === "Enter") {
+        addInputAsTag();
+        e.preventDefault();
+      }
+    };
+
+    function handleOnChange(e: React.ChangeEvent<HTMLInputElement>) {
+      setInputValue(e.target.value);
+      if (validationError) {
+        setValidationError(null);
+      }
+    }
+
+    return (
+      <div>
+        <div
+          className={cn(
+            "flex flex-wrap items-center gap-2 rounded-md border border-input bg-background p-2",
+            className
+          )}
+        >
+          {tags.map((tag, _) => (
+            <Badge
+              key={tag}
+              className="bg-transparent text-foreground border-foreground/10 hover:bg-card m-1 transition ease-in-out delay-150 hover:-translate-y-1 hover:scale-110 duration-300"
+            >
+              <span className="truncate max-w-[130px] mr-2">{tag}</span>
+              <XCircle
+                className="h-4 w-4 cursor-pointer"
+                onClick={(event) => {
+                  event.stopPropagation();
+                  removeTag(tag);
+                }}
+              />
+            </Badge>
+          ))}
+          {isPopulatedArray(tags) && <Separator />}
+          <input
```

Comment:
We can make use of react-hook-form here to 
- handle hotkeys like "Enter" automatically
- apply validating logic from zod